n3fit/src/n3fit/backends/keras_backend/multi_dense.py

```diff
@@ -0,0 +1,179 @@
+    >>> for single_model in single_models:
+    >>>     single_model.build(input_shape=(None, gridsize, features))
+    >>> test_input = tf.random.uniform(shape=(1, gridsize, features))
+    >>> multi_dense_output = multi_dense_model(test_input)
+    >>> single_dense_output = tf.stack([single_model(test_input) for single_model in single_models], axis=1)
+    >>> tf.reduce_all(tf.equal(multi_dense_output, single_dense_output))
+
+    Parameters
+    ----------
+    replica_seeds: List[int]
+        List of seeds per replica for the kernel initializer.
+    kernel_initializer: Initializer
+        Initializer class for the kernel.
+    replica_input: bool (default: True)
+        Whether the input already contains multiple replicas.
+    """
+
+    def __init__(
+        self,
+        replica_seeds: List[int],
+        kernel_initializer: Initializer,
+        replica_input: bool = True,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.replicas = len(replica_seeds)
+        self.replica_seeds = replica_seeds
+        self.kernel_initializer = MultiInitializer(
+            single_initializer=kernel_initializer, replica_seeds=replica_seeds
+        )
+        self.bias_initializer = MultiInitializer(
+            single_initializer=self.bias_initializer, replica_seeds=replica_seeds
+        )
+        self.replica_input = replica_input
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+        self.kernel = self.add_weight(
+            name="kernel",
+            shape=(self.replicas, input_dim, self.units),
+            initializer=self.kernel_initializer,
+            regularizer=self.kernel_regularizer,
+            constraint=self.kernel_constraint,
+        )
+        if self.use_bias:
+            self.bias = self.add_weight(
+                name="bias",
+                shape=(self.replicas, 1, self.units),
+                initializer=self.bias_initializer,
+                regularizer=self.bias_regularizer,
+                constraint=self.bias_constraint,
+            )
+        else:
+            self.bias = None
+        self.input_spec.axes = {-1: input_dim}
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Compute output of shape (batch_size, replicas, gridsize, units).
+
+        For the first layer, (self.replica_input is False), this is equivalent to
+        applying each replica separately and concatenating along the last axis.
+        If the input already contains multiple replica outputs, it is equivalent
+        to applying each replica to its corresponding input.
+        """
+        if inputs.dtype.base_dtype != self._compute_dtype_object.base_dtype:
+            inputs = tf.cast(inputs, dtype=self._compute_dtype_object)
+
+        input_axes = 'brnf' if self.replica_input else 'bnf'
+        einrule = input_axes + ',rfg->brng'
+        outputs = tf.einsum(einrule, inputs, self.kernel)
```

Comment:
replica_input is fixed upon initialization? Because if so it would be better to move some of this logic to the initialization part.

Also, I'm wondering whether the numerics are coming from einsum, which is different from tensordot (and the order of operations is most likely to be different).

Since the axis that would need to be passed to tensordot are known as soon as `replica_input` is known, it could be a good idea to have some logic outside that will generate the `ein_rule` and the tensordot axes rules so that it is easy to change between one and the other for

1) testing
2) benchmark (in the past `einsum` was much faster in gpu and much slower in cpu... and we don't want that)

Reply:
No I mean just matching my local master to this branch, didn't get that either.

I'm not sure it's possible to do the general case with a tensordot though, how do you implement `"brnf,rfg->brng"`?

Reply:
I'm thinking about the single replica case which is the one that needs to match.

So, you don't get the same values even if you squeeze out the extra (1,)?
We keep having these small differences and it would be good to understand where are they coming from.

Maybe is the sum rules, or maybe we need to move to float64 at least for part of the calculation.

Reply:
Ah so a flag which switches the behavior only in the single replica case? Sure, I can try a bit more to get it exactly the same. I think the differences are again just different tensorflow operations, I don't know exactly what the dense layer uses, can try to match that.

Reply:
> Ah so a flag which switches the behavior only in the single replica case?

Yes. Didn't you have that already at some point? (I remember having this conversation about eigen and tensordot already)

> I don't know exactly what the dense layer

I think it uses tensordot. But a way of isolating the issue might be to generate a dense layer and use it in that call function.

Reply:
Hmm actually there is no difference in the outputs of `MultiDense` for one replica compared to `Dense`, as I verified locally in this test, which gives exact zeroes, both after the last commit removing einsum for 1 replica, but also before.

<details>
<summary>Script</summary>

```python
import numpy as np

from tensorflow.keras.layers import Dense
from tensorflow.keras.initializers import RandomNormal
from n3fit.backends.keras_backend.multi_dense import MultiDense

replica_input = False  # check both cases

np.random.seed(1)
if replica_input:
    test_input = 10**3 * np.random.rand(1, 1, 7, 1)
else:
    test_input = 10**3 * np.random.rand(1, 7, 1)

d_ini = RandomNormal(stddev=0.1, seed=42)
d = Dense(11, kernel_initializer=d_ini)

md_ini = RandomNormal(stddev=0.1, seed=0)
md = MultiDense(replica_seeds=[42], units=11, replica_input=replica_input, kernel_initializer=md_ini)

if replica_input:
    print(d(test_input).shape, md(test_input).shape)
    print(d(test_input) - md(test_input))
else:
    print(d(test_input).shape, md(test_input)[:, 0].shape)
    print(d(test_input) - md(test_input)[:, 0])
```
</details>

Reply:
I see now that the last commit did make the difference much smaller in the CI, on the differing stop epoch in the first test, which first was about a factor 2 and now it's only ~20 epochs, about a percent. 

I did another test expanding on the previous, where I turn both layers into models and train them for a bit, and still they give identical predictions.

I don't think it's worth chasing these differences, it can be a huge rabbit hole, and we have been happy to ignore them as long as they are small several times already.

Reply:
Yes. But we have also introduced some bugs because we were not rigorous enough. Now the papers are published (and the people involved in the work need to write their respective thesis) so I'd rather be extra careful. 

If we understand that the difference is coming from dense vs multidense that's fine, we regenerate the regression tests and we continue from there... but right now we still don't know what's introducing this difference. 

Also, understanding what kind of operations screw up precision might be useful as well. We are working under the assumption that float32 is enough for the precision we claim in the PDFs, but if a change in the order of operations actually has an effect beyond 1% (which is the relative precision we are testing against) we will need to move to float64.

Reply:
I have done some more testing, printing outputs (min, max and mean) after each layer in the PDF. I see that the results are identical in the first forward pass. Then the first backward pass happens, I don't see inside that. Then, the first validation happens, where outputs start to differ, but only for the third (out of 3) dense layer, not for the first two, and not for the prefactors (but of course also for everything that depends on the third layer).

I've also repeated this with 4 rather than 3 layers, expecting to find either that it was still only the last layer, or it was layer 3 and 4 that had changed outputs. But actually now it's all layers. (all dense layers and everything that depends on them, still not the preprocessing layers ofcourse).

I've also repeated the toy example, see script below, now turning it into a model of 5 layers, all dense or all multidense, and training for 100 epochs, and I still find 0 difference.

So it's probably something in the backward pass, something to do with the multi dense layer, but that only creates a difference in the presence some of the later layers... 

If there's any other way to validate this PR, that's going to be much easier than to trace this back to exactly what's going on, I don't think I'll be able to do that.
I'm running the runcard of `231213-02-rs-nnlo-baseline-updated` now, with 120 replicas.

<details>
<summary>script</summary>

```python
import numpy as np
import tensorflow as tf

from tensorflow.keras.layers import Dense, Input, Lambda
from tensorflow.keras.initializers import RandomNormal
from n3fit.backends.keras_backend.multi_dense import MultiDense
from tensorflow.keras.models import Model

np.random.seed(1)
input_layer = Input(shape=(7, 1))
test_input = 10**3 * np.random.rand(1, 7, 1)
test_output = 10**3 * np.random.rand(1, 7, 11)

# create model with dense layers
ds = []
for i in range(5):
    d_ini = RandomNormal(stddev=0.1, seed=42+i)
    d = Dense(11, kernel_initializer=d_ini)
    ds.append(d)

ds_output = input_layer
for d in ds:
    ds_output = d(ds_output)

ds_output = Lambda(lambda nns: tf.stack(nns, axis=1), name=f"stack_replicas")([ds_output])


model_d = Model(inputs=input_layer, outputs=ds_output)
model_d.compile(optimizer='adam', loss='mse')
model_d.summary()

# create model with multi-dense layers (one replica)
mds = []
for i in range(5):
    md_ini = RandomNormal(stddev=0.1, seed=0)
    md = MultiDense(replica_seeds=[42+i], units=11, replica_input=(i != 0), kernel_initializer=md_ini)
    mds.append(md)

mds_output = input_layer
for md in mds:
    mds_output = md(mds_output)

model_md = Model(inputs=input_layer, outputs=mds_output)
model_md.compile(optimizer='adam', loss='mse')
model_md.summary()

# fit both
model_d.fit(test_input, test_output, epochs=100)
model_md.fit(test_input, test_output, epochs=100)

# compare
print(model_d(test_input) - model_md(test_input))
print(np.mean(model_d(test_input)))
print(np.mean(model_md(test_input)))

```

</summary>

Reply:
> If there's any other way to validate this PR

It's not so much validating this particular PR but rather understanding where the problem was coming from (which you just did). I want to be sure that we are not in a situation in 2/3 months from now where one of the results of the papers needs to be reproduced and it doesn't work and we don't understand why.


Could you do the exact same test but using double precision instead of single precision? That would be invaluable in this scenario because if it is a problem with numerical precision that would be a good motive to move to double for at least production runs (and the tests).
We don't need that kind of precision for things like hyperopt, but a change of a % because of an unknown change in the order of operation in the "black box" of the code is a bit worrying. If we can avoid it just by having a flag to run in double it would be great.

Thanks. And apologies for being very strict, but for now I think we need to be extremely careful ^^U

Reply:
Is there an existing flag for this? I found that for Keras one can set `tf.keras.backend.set_floatx('float64')`, which will default the weights to float64. But this creates all sorts of issues with the inputs which are all still float32.

Reply:
I added the flag in #1929, it should be working but there is an unrelated problem with the CI that I tried to fix in #1930, but apparently that wasn't it.

Reply:
Did the numerical issue go away when using `double`?

Reply:
I just checked, they became a lot less. A difference in the 10th digit in the min/max, 5th digit in the mean, of the last dense layer. Much less in earlier layers, and only in the 14th digit in the final PDF itself. (In the first validation step)
The exact same test when setting the flag to false (i.e. not using double precision) gives differences in the third digit of the min/max and second in the mean.

Reply:
Ok, then we have been too optimistic until know with doing everything in single precision.
Thanks for this. 

Is it fine if I push directly to the branch where you add the double precision flag? I will update all regression tests to use it.
I'm wondering whether we should make it default...

Reply:
(or, if you have a machine with linux in which you can re-generate the regression test you can also do it yourself ofc)

Reply:
> Is it fine if I push directly to the branch where you add the double precision flag? I will update all regression tests to use it.

Sure, no problem, I'll leave the branch alone. I don't have access to a linux machine.

> I'm wondering whether we should make it default...

I still think it's not really a problem though. I didn't mention, the difference in the final PDF, without double precision, is only in the 7th digit. No final results should be sensitive to that (in a statistically significant way), the learning process itself is much noisier than that.
Of course for reproducibility in the tests I agree it is nice to use.

Reply:
> I didn't mention, the difference in the final PDF, without double precision, is only in the 7th digit.

Ah, ok. That's very good to hear.
Then let's leave it just for tests for the time being.

Reply:
> I didn't mention, the difference in the final PDF, without double precision, is only in the 7th digit.

Just to be clear because I realized it might be interpreted differently, by final I meant after all the layers, so the normalized PDF, not the final epoch or something, still talking about the first validation step

Reply:
Ah, I see.

Well, you said you are running 231213-02-rs-nnlo-baseline-updated, that will give us the final PDF.

Reply:
Yes, something went wrong so running again, just to be sure this should be ok right?
```bash
vp-setupfit runcard-repro.yml
n3fit runcard-repro.yml 1 -r 120
evolven3fit runcard-repro 120
postfit 1 runcard-repro --at-least-nrep
```
where I took runcard-repro.yml from the fit you mention (and adapted it to run in parallel, have to set same_trvl_per_replica to true though).

Reply:
No, don´t run like this `n3fit runcard-repro.yml 1 -r 120`
Run it one replica at a time (we want to compare with the fit that @RoyStegeman did)

Also, for postfit you would want 100 replicas.

Reply:
Hmm but that wouldnt test the multi replica implementation. Should we perhaps get #1788 merged first and then do the above test with same_trvl_per_replica True?

Reply:
Alright, running them. Should I use `evolven3fit_new`? I thought that had become the regular `evolven3fit` but the command is not recognised. Perhaps I need to reinstall again?

Reply:
I left it as `evolven3fit_new` https://github.com/NNPDF/nnpdf/blob/e4a19925e06c93002956bac498338a1123ef2964/pyproject.toml#L43C1-L43C16 since I assumed people would have it already as part of their scripts with that name.

Reply:
Here is the [comparison](https://vp.nnpdf.science/FGwCRitnQhmqBYBjeQWS7Q==/). (Perhaps some parts are missing, I wasn't sure if the comparison really completed or quit early, if so let me know and I'll rerun it)

Reply:
@scarlehoff We just had a look at the report during the meeting, and it looked good to us. If you agree and merge #1929 I can rebase and regenerate the regression tests.

Reply:
Yes, please!

If you cannot regenerate the tests I can do so.

Reply:
Well, actually, go on without regenerating the tests.
Better to finish the PR as if they were working and we update them only at the end!